Kuadrant/limitador-operator
# Limits in limitador

Closes https://github.com/Kuadrant/limitador-operator/issues/21
Closes https://github.com/Kuadrant/limitador-operator/issues/27

This PR gets rid of the `RateLimit` CRD and includes the limits definition within the `Limitador` one. It also removes the code in charge of making the HTTP requests to the _Limitador_ service that _used_ to expose an API to manage limits.

It also inculdes https://github.com/Kuadrant/limitador-operator/pull/28 `ConfigMap` that holds the _state_ and _limits_ (configuration needed for setting up _Limitador_) which it's reconciled by the _limitador-controller_

The `ConfigMap` looks like:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app: limitador
  name: limits-config-limitador-1 # Name constructed from "limits-config-" prefix and the `Limitador` CR name
data:
  hash: 32trgresg3ydsagas # hash built from the actual limits. Used for reconciling (state)
  limitador-config.yaml: |
  limits:
  - conditions: ["get-toy == yes"]
    max_value: 2
    namespace: toystore-app
    seconds: 30
    variables: []
  - conditions:
    - "admin == yes"
    max_value: 2
    namespace: toystore-app
    seconds: 30
    variables: []
  - conditions: ["vhaction == yes"]
    max_value: 6
    namespace: toystore-app
    seconds: 30
    variables: []            
```                                

And the file it's fed to the deployment env varialbe `LIMITS_FILE` and mounted as a _Volume_ to it`s deployment.

**Verification steps:**

1. Lunch the cluster setup with `make local-setup`
2. When ready, create a new namespace like `kubectl create namespace "limitador-test"`
3. Then, apply the following Limitador CR:
```bash
kubectl -n limitador-test apply -f -<<EOF
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador
metadata:
  name: limitador
spec:
  replicas: 1
  version: "0.4.0"
  listener:
    http:
      port: 8080
    grpc:
      port: 8081
  limits:
    - conditions: ["get-toy == yes"]
      max_value: 2
      namespace: toystore-app
      seconds: 30
      variables: []
EOF

```
4. Then between the couple of things that got crated by the operator, you should see the following `ConfigMap`:
```bash
kubectl -n limitador-test get cm/limits-config-limitador -o yaml
```
```yaml
apiVersion: v1
data:
  hash: c7fabb796b22ffef0a4207da931d67bd
  limitador-config.yaml: |
    - conditions:
      - get-toy == yes
      max_value: 2
      namespace: toystore-app
      seconds: 30
      variables: []
kind: ConfigMap
metadata:
  creationTimestamp: "2022-06-27T16:27:20Z"
  labels:
    app: limitador
  name: limits-config-limitador
  namespace: limitador-test
  resourceVersion: "1311"
  uid: a326cea3-80ba-4ef4-976f-aa75f9e7157c
```
 
5 Check if the `limitador-config.yaml` was successfully mounted on the deployed _Limitador_ pod
```bash
kubectl exec -ti $(kubectl get pods -n limitador-test | grep 'limitador' | cut -d " " -f1) -n limitador-test -- /bin/sh
```

Once inside the pod:

```bash
cat /home/limitador/etc/limitador-config.yaml
```

And you should see:
```yaml
- conditions:
  - get-toy == yes
  max_value: 20
  namespace: toystore-app
  seconds: 30
  variables: []
```

6. Checking the Update of the `Limitador`CR:
```bash
kubectl -n limitador-test apply -f -<<EOF
apiVersion: limitador.kuadrant.io/v1alpha1
kind: Limitador
metadata:
  name: limitador
spec:
  replicas: 1
  version: "0.4.0"
  listener:
    http:
      port: 8080
    grpc:
      port: 8081
  limits:
    - conditions: ["get-toy == yes"]
      max_value: 2
      namespace: toystore-app
      seconds: 30
      variables: []
    - conditions:
      - "admin == yes"
      max_value: 2
      namespace: toystore-app
      seconds: 30
      variables: []
EOF
```
7. Now the `ConfigMap` should be updated too:
```bash
kubectl -n limitador-test get cm/limits-config-limitador -o yaml
```

```yaml
apiVersion: v1
data:
  hash: a230fdfb0d8665ef8e2e61844c14c106
  limitador-config.yaml: |
    - conditions:
      - get-toy == yes
      max_value: 2
      namespace: toystore-app
      seconds: 30
      variables: []
    - conditions:
      - admin == yes
      max_value: 2
      namespace: toystore-app
      seconds: 30
      variables: []
kind: ConfigMap
metadata:
  creationTimestamp: "2022-06-28T14:42:32Z"
  labels:
    app: limitador
  name: limits-config-limitador
  namespace: limitador-test
  resourceVersion: "37131"
  uid: bde21563-c1ae-4381-b48d-3bfb4a79fd25
```

8. And if you check the config file within the _Limitador_ instance:
```bash
kubectl exec -ti $(kubectl get pods -n limitador-test | grep 'limitador' | cut -d " " -f1) -n limitador-test -- /bin/sh
```
```bash
cat /home/limitador/etc/limitador-config.yaml
```
```yaml
- conditions:
  - get-toy == yes
  max_value: 2
  namespace: toystore-app
  seconds: 30
  variables: []
- conditions:
  - admin == yes
  max_value: 2
  namespace: toystore-app
  seconds: 30
  variables: []
```
